src/folly/FollyKeepaliveTimer.cpp

```diff
@@ -27,8 +27,18 @@ void FollyKeepaliveTimer::schedule() {
   eventBase_.runAfterDelay(
       [this, running]() {
         if (*running) {
-          connection_->sendKeepalive();
-          schedule();
+          if (pending_) {
+            stop();
+
+            connection_->outputFrameOrEnqueue(
```

Comment:
I am not a big fan on spreading the behavior of failed keepalive over here.
Do you mind creating a new method ConnectionAutomaton::keepaliveFailed and putting the {outputFrameOrEnqueue, disconnect} logic there?